titus-ext/zookeeper/src/main/java/com/netflix/titus/ext/zookeeper/supervisor/ZookeeperLeaderElector.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.BackgroundCallback;
+import org.apache.curator.framework.api.CuratorEvent;
+import org.apache.curator.framework.recipes.leader.LeaderLatch;
+import org.apache.curator.framework.recipes.leader.LeaderLatchListener;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+import rx.subjects.PublishSubject;
+
+import static org.apache.zookeeper.KeeperException.Code.OK;
+
+@Singleton
+public class ZookeeperLeaderElector implements LeaderElector {
+    private static final Logger logger = LoggerFactory.getLogger(ZookeeperLeaderElector.class);
+
+    private final LeaderActivator leaderActivator;
+    private final ZookeeperPaths zookeeperPaths;
+    private final ObjectMapper jsonMapper;
+    private final MasterDescription masterDescription;
+    private final CuratorFramework curator;
+    // The path where a selected leader announces itself.
+    private final String leaderPath;
+
+    private final PublishSubject<MasterState> electionSubject = PublishSubject.create();
+    private final Observable<MasterState> electionObserver = ObservableExt.protectFromMissingExceptionHandlers(electionSubject, logger);
+
+    private final AtomicReference<LeaderElectionProcess> leaderElectionProcessRef = new AtomicReference<>();
+
+    @Inject
+    public ZookeeperLeaderElector(LeaderActivator leaderActivator,
+                                  CuratorService curatorService,
+                                  ZookeeperPaths zookeeperPaths,
+                                  MasterDescription masterDescription) {
+        this.leaderActivator = leaderActivator;
+        this.zookeeperPaths = zookeeperPaths;
+        this.jsonMapper = ObjectMappers.defaultMapper();
+        this.masterDescription = masterDescription;
+        this.curator = curatorService.getCurator();
+        this.leaderPath = zookeeperPaths.getLeaderAnnouncementPath();
+
+        initialize();
+    }
+
+    private void initialize() {
+        try {
+            Stat pathStat = curator.checkExists().forPath(leaderPath);
+            // Create the path only if the path does not exist
+            if (pathStat == null) {
+                curator.create()
+                        .creatingParentsIfNeeded()
+                        .withMode(CreateMode.PERSISTENT)
+                        .forPath(leaderPath);
+            }
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to initialize leader elector path: " + e.getMessage(), e);
+        }
+    }
+
+    @PreDestroy
+    public void shutdown() {
+        LeaderElectionProcess process = leaderElectionProcessRef.getAndSet(null);
+        if (process != null) {
+            process.close();
+        }
+    }
+
+    @Override
+    public boolean join() {
+        synchronized (leaderElectionProcessRef) {
```

Comment:
Because we do not want to create excessive `LeaderElectionProcess` instances. This could happen in the case above.

Reply:
The problem is that as it is, it can race with both `shutdown()`, and (more importantly) `leaveIfNotLeader()`. In the latter case, both `join()` and `leaveIfNotLeader()` are running concurrently, and both detect that `leaderElectionProcessRef.get() == null`, we may end up leaking a `LeaderLatch`.

Either we need to refactor this, or we need to serialize the other uses of `leaderElectionProcessRef` in `shutdown()` and `leaveIfNotLeader` too.

Reply:
hrmm, looking again, I'm wrong. This is fine and there are no races.